dask/dataframe/core.py

```diff
@@ -357,6 +357,39 @@ def _rebuild(self, dsk, *, rename=None):
     def _constructor(self):
         return new_dd_object
 
+    @property
+    def divisions(self):
+        """
+        Tuple of ``npartitions + 1`` values, in ascending order, marking the
+        lower/upper bounds of each partition's index. Divisions allow Dask
+        to know which partition will contain a given value, significantly
+        speeding up operations like `loc`, `merge`, and `groupby` by not
+        having to search the full dataset.
+
+        Example: for ``divisions = (0, 10, 50, 100)``, there are three partitions,
+        where the index in each partition contains values [0, 10), [10, 50),
+        and [50, 100), respectively. Dask therefore knows ``df.loc[45]``
+        will be in the second partition.
+
+        When every item in ``divisions`` is ``None``, the divisions are unknown.
+        Most operations can still be performed, but some will be much slower,
+        and a few may fail.
+
+        It is uncommon to set ``divisions`` directly. Instead, use ``set_index``,
+        which sorts and splits the data as needed.
+        See https://docs.dask.org/en/latest/dataframe-design.html#partitions.
+        """
+        return self._divisions
+
+    @divisions.setter
+    def divisions(self, value):
+        if None in value:
```

Comment:
Can we also check somewhere that `len(value) == self.npartitions + 1`?

Reply:
Since `npartitions` is based off of the length of divisions, could we just check that `len(value) == len(self._divisions)`?

Reply:
@jrbourbeau brought up it might be less surprising, and allow us to do a deprecation cycle, if instead of immediately coercing to tuple, we warn if divisions isn't a tuple. This will have the added benefit of snuffling out any places in the dask codebase where we are setting divisions with a list.

Reply:
Yeah, specifically I'll propose we:

1. Emit a deprecation warning when the divisions setter encounters a non-`tuple` `value` that says something like "we will automatically convert this to a tuple in the future". This will let us inform users who are explicitly setting `divisions` today about this change. This may be overkill, but is in line with our desire to be better about informing users of behavior changes. This will also cause errors to be raised in our CI in all the places internally where `divisions` is a non-`tuple` today.
2. Change all the places internally in `dask/dask` where we're not using `tuple` divisions to be a ` tuple`. This should result in CI passing again.
3. Open an issue about the deprecation and when we the deprecation cycle should end

Reply:
That sounds good to me - in that case, would we also like to hold off on additional checks on the length / contexts of `value` until that deprecation cycle ends?

Reply:
Good point. Checks on length, monotonically increasing, etc. should be able valid for both `tuple`s and `list`s, so we should be okay to include them

Reply:
Also, if we have a deprecation cycle for this and are able to resolve it completely in upstream Dask, would it make sense to simply raise a `ValueError` when non-tuple divisions are provided in the future? This obviously comes with the potential for downstream breakage, but might make it the reasoning for it more obvious (e.g. a downstream test breaking because it attempts to set list divisions vs. later on when it tries to assert that divisions is a list).

Reply:
My sense is a deprecation cycle is still better than raising an error in that case. The goal would be to avoid downstream breakages without some kind of warning first.

Reply:
Yeah, definitely in favor of the deprecation cycle here since there are known downstream libraries this would impact (dask-cudf and dask-sql in my case); I just prefer shifting from a `DeprecationWarning` to a `ValueError` in the future rather than implicitly converting to a tuple

Reply:
Ah, I see. I misinterpreted your previous comment. Post-deprecation, raising an error (probably a `TypeError` in this case) seems like a reasonable thing to do -- no strong opinion from me

Reply:
Following up on the original concern here (length check on the new `divisions`), it looks like there are several places in the codebase where a `divisions` with a different length will be provided as a means of repartitioning the dataframe. In cases like this, would we prefer instead using `df.repartition(divisions=...)`?

I can think of cases where this currently wouldn't work (for example, if we want to repartition using unknown divisions) that seem to motivate leaving out the length check for now, unless we want to address those areas in this PR

Reply:
I see how that could be a chicken and the egg problem :), but I don't understand how setting divisions to a different length would work for repartitioning the dataframe. Do you have a specific example of where that happens?